lib/vrl/stdlib/src/parse_nginx_log.rs

```diff
@@ -0,0 +1,225 @@
+                result: Ok(
+                    r#"s'{"client":"172.17.0.1","timestamp":"2021-03-31T12:04:07Z","request":"GET / HTTP/1.1","method":"GET","path":"/","protocol":"HTTP/1.0","status":200,"size":612,"agent":"curl/7.75.0"}'"#,
+                ),
+            },
+            Example {
+                title: "parse nginx error log",
+                source: r#"encode_json(parse_nginx_log!(s'2021/04/01 13:02:31 [error] 31#31: *1 open() "/usr/share/nginx/html/not-found" failed (2: No such file or directory), client: 172.17.0.1, server: localhost, request: "POST /not-found HTTP/1.1", host: "localhost:8081"', "error"))"#,
+                result: Ok(
+                    r#"s'{"timestamp":"2021-04-01T13:02:31Z","severity":"error","pid":31,"tid":31,"cid":1,"message":"open() "/usr/share/nginx/html/not-found" failed (2: No such file or directory)","client":"172.17.0.1","server":"localhost","request":"POST /not-found HTTP/1.1","host":"localhost","port":"8081"}'"#,
+                ),
+            },
+        ]
+    }
+}
+
+fn regex_for_format(format: &[u8]) -> &Regex {
+    match format {
+        b"combined" => &*log_util::REGEX_NGINX_COMBINED_LOG,
+        b"error" => &*log_util::REGEX_NGINX_ERROR_LOG,
+        _ => unreachable!(),
+    }
+}
+
+fn time_format_for_format(format: &[u8]) -> String {
+    match format {
+        b"combined" => "%d/%b/%Y:%T %z".to_owned(),
+        b"error" => "%Y/%m/%d %H:%M:%S".to_owned(),
+        _ => unreachable!(),
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseNginxLogFn {
+    value: Box<dyn Expression>,
+    format: Bytes,
+    timestamp_format: Option<Box<dyn Expression>>,
+}
+
+impl Expression for ParseNginxLogFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let bytes = self.value.resolve(ctx)?;
+        let message = bytes.try_bytes_utf8_lossy()?;
+        let timestamp_format = match &self.timestamp_format {
+            None => time_format_for_format(self.format.as_ref()),
+            Some(timestamp_format) => timestamp_format
+                .resolve(ctx)?
+                .try_bytes_utf8_lossy()?
+                .to_string(),
+        };
+
+        let regex = regex_for_format(self.format.as_ref());
+
+        let captures = regex.captures(&message).ok_or("failed parsing log line")?;
+
+        log_util::log_fields(&regex, &captures, &timestamp_format).map_err(Into::into)
+    }
+
+    fn type_def(&self, _: &state::Compiler) -> TypeDef {
+        TypeDef::new()
+            .fallible()
+            .object(match self.format.as_ref() {
+                b"combined" => type_def_combined(),
+                b"error" => type_def_error(),
+                _ => unreachable!(),
+            })
+    }
+}
+
+fn type_def_combined() -> BTreeMap<&'static str, TypeDef> {
+    map! {
+         "client": Kind::Bytes | Kind::Null,
```

Comment:
From what I can gather there will always be a "client" field returned (I may be mistaken..). So this type could be: 

```
"client": Kind::Bytes,
```

Same with the other fields.